File: c/CSharp/SymbolTable.cs
using System;
using System.Collections.Generic;
using System.Text;

public class SymbolTable {
    private Stack<Symbol> scopeStack = new Stack<Symbol>();
    private int blockCounter = 0;

    public SymbolTable() {
        var globalScope = new Symbol() { Name = "global", Classification = new HashSet<TypeClassification>() {TypeClassification.Global_} };
        scopeStack.Push(globalScope);

        Define(new Symbol() { Name = "auto", Classification = new HashSet<TypeClassification>(){TypeClassification.StorageClassSpecifier_}, Predefined = true });
        Define(new Symbol() { Name = "constexpr", Classification = new HashSet<TypeClassification>() { TypeClassification.StorageClassSpecifier_ }, Predefined = true });
        Define(new Symbol() { Name = "extern", Classification = new HashSet<TypeClassification>() { TypeClassification.StorageClassSpecifier_ }, Predefined = true });
        Define(new Symbol() { Name = "register", Classification = new HashSet<TypeClassification>() { TypeClassification.StorageClassSpecifier_ }, Predefined = true });
        Define(new Symbol() { Name = "static", Classification = new HashSet<TypeClassification>() { TypeClassification.StorageClassSpecifier_ }, Predefined = true });
        Define(new Symbol() { Name = "thread_local", Classification = new HashSet<TypeClassification>() { TypeClassification.StorageClassSpecifier_ }, Predefined = true });
        Define(new Symbol() { Name = "_Thread_local", Classification = new HashSet<TypeClassification>() { TypeClassification.StorageClassSpecifier_ }, Predefined = true });
        Define(new Symbol() { Name = "typedef", Classification = new HashSet<TypeClassification>() { TypeClassification.StorageClassSpecifier_ }, Predefined = true });

        Define(new Symbol() { Name = "enum", Classification = new HashSet<TypeClassification>() { TypeClassification.EnumSpecifier_ }, Predefined = true });

        Define(new Symbol() { Name = "struct", Classification = new HashSet<TypeClassification>() { TypeClassification.StorageClassSpecifier_ }, Predefined = true });
        Define(new Symbol() { Name = "union", Classification = new HashSet<TypeClassification>() { TypeClassification.StorageClassSpecifier_ }, Predefined = true });

        Define(new Symbol() { Name = "const", Classification = new HashSet<TypeClassification>() { TypeClassification.TypeQualifier_ }, Predefined = true });
        Define(new Symbol() { Name = "restrict", Classification = new HashSet<TypeClassification>() { TypeClassification.TypeQualifier_ }, Predefined = true });
        Define(new Symbol() { Name = "volatile", Classification = new HashSet<TypeClassification>() { TypeClassification.TypeQualifier_ }, Predefined = true });
        Define(new Symbol() { Name = "_Atomic", Classification = new HashSet<TypeClassification>() { TypeClassification.TypeQualifier_, TypeClassification.AtomicTypeSpecifier_}, Predefined = true });

        Define(new Symbol() { Name = "void", Classification = new HashSet<TypeClassification>() { TypeClassification.TypeSpecifier_ }, Predefined = true });
        Define(new Symbol() { Name = "char", Classification = new HashSet<TypeClassification>() { TypeClassification.TypeSpecifier_ }, Predefined = true });
        Define(new Symbol() { Name = "short", Classification = new HashSet<TypeClassification>() { TypeClassification.TypeSpecifier_ }, Predefined = true });
        Define(new Symbol() { Name = "int", Classification = new HashSet<TypeClassification>() { TypeClassification.TypeSpecifier_ }, Predefined = true });
        Define(new Symbol() { Name = "long", Classification = new HashSet<TypeClassification>() { TypeClassification.TypeSpecifier_ }, Predefined = true });
        Define(new Symbol() { Name = "float", Classification = new HashSet<TypeClassification>() { TypeClassification.TypeSpecifier_ }, Predefined = true });
        Define(new Symbol() { Name = "double", Classification = new HashSet<TypeClassification>() { TypeClassification.TypeSpecifier_ }, Predefined = true });
        Define(new Symbol() { Name = "signed", Classification = new HashSet<TypeClassification>() { TypeClassification.TypeSpecifier_ }, Predefined = true });
        Define(new Symbol() { Name = "unsigned", Classification = new HashSet<TypeClassification>() { TypeClassification.TypeSpecifier_ }, Predefined = true });
        Define(new Symbol() { Name = "_BitInt", Classification = new HashSet<TypeClassification>() { TypeClassification.TypeSpecifier_ }, Predefined = true });
        Define(new Symbol() { Name = "bool", Classification = new HashSet<TypeClassification>() { TypeClassification.TypeSpecifier_ }, Predefined = true });
        Define(new Symbol() { Name = "_Bool", Classification = new HashSet<TypeClassification>() { TypeClassification.TypeSpecifier_ }, Predefined = true });
        Define(new Symbol() { Name = "_Complex", Classification = new HashSet<TypeClassification>() { TypeClassification.TypeSpecifier_ }, Predefined = true });
        Define(new Symbol() { Name = "_Decimal32", Classification = new HashSet<TypeClassification>() { TypeClassification.TypeSpecifier_ }, Predefined = true });
        Define(new Symbol() { Name = "_Decimal64", Classification = new HashSet<TypeClassification>() { TypeClassification.TypeSpecifier_ }, Predefined = true });
        Define(new Symbol() { Name = "_Decimal128", Classification = new HashSet<TypeClassification>() { TypeClassification.TypeSpecifier_ }, Predefined = true });
        Define(new Symbol() { Name = "__m128", Classification = new HashSet<TypeClassification>() { TypeClassification.TypeSpecifier_ }, Predefined = true });
        Define(new Symbol() { Name = "__m128d", Classification = new HashSet<TypeClassification>() { TypeClassification.TypeSpecifier_ }, Predefined = true });
        Define(new Symbol() { Name = "__m128i", Classification = new HashSet<TypeClassification>() { TypeClassification.TypeSpecifier_ }, Predefined = true });
        Define(new Symbol() { Name = "__extension__", Classification = new HashSet<TypeClassification>() { TypeClassification.TypeSpecifier_ }, Predefined = true });

        Define(new Symbol() { Name = "__builtin_va_list", Classification = new HashSet<TypeClassification>() { TypeClassification.TypeSpecifier_ }, Predefined = true });
        Define(new Symbol() { Name = "__builtin_has_attribute", Classification = new HashSet<TypeClassification>() { TypeClassification.TypeSpecifier_ }, Predefined = true });
        Define(new Symbol() { Name = "__builtin_speculation_safe_value", Classification = new HashSet<TypeClassification>() { TypeClassification.TypeSpecifier_ }, Predefined = true });
        Define(new Symbol() { Name = "__builtin_types_compatible_p", Classification = new HashSet<TypeClassification>() { TypeClassification.TypeSpecifier_ }, Predefined = true });
        Define(new Symbol() { Name = "__builtin_choose_expr", Classification = new HashSet<TypeClassification>() { TypeClassification.TypeSpecifier_ }, Predefined = true });
        Define(new Symbol() { Name = "__builtin_tgmath", Classification = new HashSet<TypeClassification>() { TypeClassification.TypeSpecifier_ }, Predefined = true });
        Define(new Symbol() { Name = "__builtin_constant_p", Classification = new HashSet<TypeClassification>() { TypeClassification.TypeSpecifier_ }, Predefined = true });
        Define(new Symbol() { Name = "__builtin_is_constant_evaluated", Classification = new HashSet<TypeClassification>() { TypeClassification.TypeSpecifier_ }, Predefined = true });
        Define(new Symbol() { Name = "__builtin_bit_cast", Classification = new HashSet<TypeClassification>() { TypeClassification.TypeSpecifier_ }, Predefined = true });
        Define(new Symbol() { Name = "__builtin_expect", Classification = new HashSet<TypeClassification>() { TypeClassification.TypeSpecifier_ }, Predefined = true });
        Define(new Symbol() { Name = "__builtin_expect_with_probability", Classification = new HashSet<TypeClassification>() { TypeClassification.TypeSpecifier_ }, Predefined = true });
        Define(new Symbol() { Name = "__builtin_trap", Classification = new HashSet<TypeClassification>() { TypeClassification.TypeSpecifier_ }, Predefined = true });
//        Define(new Symbol() { Name = "__builtin_unreachable", Classification = new HashSet<TypeClassification>() { TypeClassification.FunctionSpecifier_ }, Predefined = true });
        Define(new Symbol() { Name = "__builtin_assoc_barrier", Classification = new HashSet<TypeClassification>() { TypeClassification.TypeSpecifier_ }, Predefined = true });
        Define(new Symbol() { Name = "__builtin_assume_aligned", Classification = new HashSet<TypeClassification>() { TypeClassification.TypeSpecifier_ }, Predefined = true });
        Define(new Symbol() { Name = "__builtin_LINE", Classification = new HashSet<TypeClassification>() { TypeClassification.TypeSpecifier_ }, Predefined = true });
        Define(new Symbol() { Name = "__builtin_FUNCTION", Classification = new HashSet<TypeClassification>() { TypeClassification.TypeSpecifier_ }, Predefined = true });
        Define(new Symbol() { Name = "__builtin_FILE", Classification = new HashSet<TypeClassification>() { TypeClassification.TypeSpecifier_ }, Predefined = true });
        Define(new Symbol() { Name = "__builtin___clear_cache", Classification = new HashSet<TypeClassification>() { TypeClassification.TypeSpecifier_ }, Predefined = true });
        Define(new Symbol() { Name = "__builtin_prefetch", Classification = new HashSet<TypeClassification>() { TypeClassification.Function_ }, Predefined = true });
        Define(new Symbol() { Name = "__builtin_classify_type", Classification = new HashSet<TypeClassification>() { TypeClassification.TypeSpecifier_ }, Predefined = true });
        Define(new Symbol() { Name = "__builtin_extend_pointer", Classification = new HashSet<TypeClassification>() { TypeClassification.TypeSpecifier_ }, Predefined = true });
        Define(new Symbol() { Name = "__builtin_goacc_parlevel_id", Classification = new HashSet<TypeClassification>() { TypeClassification.TypeSpecifier_ }, Predefined = true });
        Define(new Symbol() { Name = "__builtin_goacc_parlevel_size", Classification = new HashSet<TypeClassification>() { TypeClassification.TypeSpecifier_ }, Predefined = true });

        Define(new Symbol() { Name = "inline", Classification = new HashSet<TypeClassification>() { TypeClassification.FunctionSpecifier_ }, Predefined = true });
        Define(new Symbol() { Name = "_Noreturn", Classification = new HashSet<TypeClassification>() { TypeClassification.FunctionSpecifier_ }, Predefined = true });
        Define(new Symbol() { Name = "__inline__", Classification = new HashSet<TypeClassification>() { TypeClassification.FunctionSpecifier_ }, Predefined = true });
        
        Define(new Symbol() { Name = "__cdecl", Classification = new HashSet<TypeClassification>() { TypeClassification.FunctionSpecifier_ }, Predefined = true }); // MS
        Define(new Symbol() { Name = "__clrcall", Classification = new HashSet<TypeClassification>() { TypeClassification.FunctionSpecifier_ }, Predefined = true }); // MS
        Define(new Symbol() { Name = "__stdcall", Classification = new HashSet<TypeClassification>() { TypeClassification.FunctionSpecifier_ }, Predefined = true }); // MS
        Define(new Symbol() { Name = "__fastcall", Classification = new HashSet<TypeClassification>() { TypeClassification.FunctionSpecifier_ }, Predefined = true }); // MS
        Define(new Symbol() { Name = "__thiscall", Classification = new HashSet<TypeClassification>() { TypeClassification.FunctionSpecifier_ }, Predefined = true }); // MS
        Define(new Symbol() { Name = "__vectorcall", Classification = new HashSet<TypeClassification>() { TypeClassification.FunctionSpecifier_ }, Predefined = true }); // MS

        Define(new Symbol() { Name = "_purecall", Classification = new HashSet<TypeClassification>() { TypeClassification.TypeSpecifier_ }, Predefined = true }); // gcc
        Define(new Symbol() { Name = "_purecall_handler", Classification = new HashSet<TypeClassification>() { TypeClassification.TypeSpecifier_ }, Predefined = true }); // gcc
        Define(new Symbol() { Name = "_onexit_t", Classification = new HashSet<TypeClassification>() { TypeClassification.TypeSpecifier_ }, Predefined = true }); // gcc
        Define(new Symbol() { Name = "_locale_t", Classification = new HashSet<TypeClassification>() { TypeClassification.TypeSpecifier_ }, Predefined = true }); // gcc
        Define(new Symbol() { Name = "_invalid_parameter_handler", Classification = new HashSet<TypeClassification>() { TypeClassification.TypeSpecifier_ }, Predefined = true }); // gcc
        Define(new Symbol() { Name = "__inline", Classification = new HashSet<TypeClassification>() { TypeClassification.TypeSpecifier_ }, Predefined = true }); // gcc
        

        Define(new Symbol() { Name = "__int8", Classification = new HashSet<TypeClassification>() { TypeClassification.TypeSpecifier_ }, Predefined = true }); // gcc
        Define(new Symbol() { Name = "__int16", Classification = new HashSet<TypeClassification>() { TypeClassification.TypeSpecifier_ }, Predefined = true }); // gcc
        Define(new Symbol() { Name = "__int32", Classification = new HashSet<TypeClassification>() { TypeClassification.TypeSpecifier_ }, Predefined = true }); // gcc
        Define(new Symbol() { Name = "__int64", Classification = new HashSet<TypeClassification>() { TypeClassification.TypeSpecifier_ }, Predefined = true }); // gcc
        Define(new Symbol() { Name = "__int128", Classification = new HashSet<TypeClassification>() { TypeClassification.TypeSpecifier_ }, Predefined = true }); // gcc
        Define(new Symbol() { Name = "_Float16", Classification = new HashSet<TypeClassification>() { TypeClassification.TypeSpecifier_ }, Predefined = true }); // gcc
        Define(new Symbol() { Name = "_Float32", Classification = new HashSet<TypeClassification>() { TypeClassification.TypeSpecifier_ }, Predefined = true }); // gcc
        Define(new Symbol() { Name = "_Float64", Classification = new HashSet<TypeClassification>() { TypeClassification.TypeSpecifier_ }, Predefined = true }); // gcc
        Define(new Symbol() { Name = "_Float128", Classification = new HashSet<TypeClassification>() { TypeClassification.TypeSpecifier_ }, Predefined = true }); // gcc
        Define(new Symbol() { Name = "__v8hf", Classification = new HashSet<TypeClassification>() { TypeClassification.TypeSpecifier_ }, Predefined = true }); // gcc
        Define(new Symbol() { Name = "__bf16", Classification = new HashSet<TypeClassification>() { TypeClassification.TypeSpecifier_ }, Predefined = true }); // gcc
        Define(new Symbol() { Name = "__v16bf", Classification = new HashSet<TypeClassification>() { TypeClassification.TypeSpecifier_ }, Predefined = true }); // gcc
        
        Define(new Symbol() { Name = "__declspec", Classification = new HashSet<TypeClassification>() { TypeClassification.FunctionSpecifier_ }, Predefined = true }); // MS
        Define(new Symbol() { Name = "__attribute__", Classification = new HashSet<TypeClassification>() { TypeClassification.FunctionSpecifier_ }, Predefined = true }); // GCC

        Define(new Symbol() { Name = "alignas", Classification = new HashSet<TypeClassification>() { TypeClassification.AlignmentSpecifier_ }, Predefined = true });
        Define(new Symbol() { Name = "align", Classification = new HashSet<TypeClassification>() { TypeClassification.AlignmentSpecifier_ }, Predefined = true });
    }

    public void EnterScope(Symbol newScope) {
        var current = scopeStack.Peek();
        if (newScope == current) return;
        scopeStack.Push(newScope);
    }

    public void ExitScope() {
        var current = scopeStack.Peek();
        scopeStack.Pop();
        if (scopeStack.Count == 0)
            throw new Exception();
    }

    public Symbol CurrentScope()
    {
        if (scopeStack.Count == 0) return null;
        var current_scope = scopeStack.Peek();
        return current_scope;
    }
    
    public bool Define(Symbol symbol) {
        var currentScope = CurrentScope();
        return this.DefineInScope(currentScope, symbol);
    }

    public bool DefineInScope(Symbol currentScope, Symbol symbol) {
        if (currentScope.Members.ContainsKey(symbol.Name)) {
            return false; // Symbol already defined in the current scope
        }
        symbol.Parent = currentScope;
        currentScope.Members[symbol.Name] = symbol;
        return true;
    }

    public Symbol Resolve(string name, Symbol start_scope = null)
    {
        if (start_scope == null)
        {
            // Stack iteration in C# is already from top (innermost) to bottom (outermost)
            foreach (Symbol scope in scopeStack)
            {
                if (scope.Members.TryGetValue(name, out Symbol symbol))
                {
                    return symbol;
                }
            }

            return null; // Symbol not found
        }
        else
        {
            if (start_scope.Members.TryGetValue(name, out Symbol symbol))
            {
                return symbol;
            }

            return null; // Symbol not found
        }
    }

    public Symbol PushBlockScope()
    {
        var blockScope = new Symbol()
        {
            Name = "block" + (++blockCounter),
            Classification = new HashSet<TypeClassification>() { TypeClassification.Block_ },
            Predefined = true
        };
        EnterScope(blockScope);
        return blockScope;
    }

    public void PopBlockScope()
    {
        ExitScope();
    }

    public override string ToString()
    {
        StringBuilder sb = new StringBuilder();
        // Get the global scope (bottom of stack)
        var scopes = scopeStack.ToArray();
        if (scopes.Length > 0)
        {
            ToStringHelper(sb, scopes[scopes.Length - 1], 0);
        }
        return sb.ToString();
    }

    private void ToStringHelper(StringBuilder sb, Symbol scope, int depth)
    {
        string indent = new string(' ', depth * 2);
        foreach (var entry in scope.Members)
        {
            var sym = entry.Value;
            if (!sym.Predefined)
            {
                sb.AppendLine(indent + sym.ToString());
            }
            // Recursively print nested scopes
            if (sym.Classification.Contains(TypeClassification.Block_) ||
                sym.Classification.Contains(TypeClassification.Function_))
            {
                ToStringHelper(sb, sym, depth + 1);
            }
        }
    }
}

File: c/Dart/SymbolTable.dart
import 'Symbol.dart';
import 'TypeClassification.dart';

class SymbolTable {
  final List<Symbol> _scopeStack = [];
  int _blockCounter = 0;

  SymbolTable() {
    var globalScope = _createSymbol("global", TypeClassification.global);
    _scopeStack.add(globalScope);

    define(_createSymbol("auto", TypeClassification.storageClassSpecifier));
    define(_createSymbol("constexpr", TypeClassification.storageClassSpecifier));
    define(_createSymbol("extern", TypeClassification.storageClassSpecifier));
    define(_createSymbol("register", TypeClassification.storageClassSpecifier));
    define(_createSymbol("static", TypeClassification.storageClassSpecifier));
    define(_createSymbol("thread_local", TypeClassification.storageClassSpecifier));
    define(_createSymbol("_Thread_local", TypeClassification.storageClassSpecifier));
    define(_createSymbol("typedef", TypeClassification.storageClassSpecifier));

    define(_createSymbol("enum", TypeClassification.enumSpecifier));

    define(_createSymbol("struct", TypeClassification.storageClassSpecifier));
    define(_createSymbol("union", TypeClassification.storageClassSpecifier));

    define(_createSymbol("const", TypeClassification.typeQualifier));
    define(_createSymbol("restrict", TypeClassification.typeQualifier));
    define(_createSymbol("volatile", TypeClassification.typeQualifier));
    define(_createSymbolMulti("_Atomic", [TypeClassification.typeQualifier, TypeClassification.atomicTypeSpecifier]));

    define(_createSymbol("void", TypeClassification.typeSpecifier));
    define(_createSymbol("char", TypeClassification.typeSpecifier));
    define(_createSymbol("short", TypeClassification.typeSpecifier));
    define(_createSymbol("int", TypeClassification.typeSpecifier));
    define(_createSymbol("long", TypeClassification.typeSpecifier));
    define(_createSymbol("float", TypeClassification.typeSpecifier));
    define(_createSymbol("double", TypeClassification.typeSpecifier));
    define(_createSymbol("signed", TypeClassification.typeSpecifier));
    define(_createSymbol("unsigned", TypeClassification.typeSpecifier));
    define(_createSymbol("_BitInt", TypeClassification.typeSpecifier));
    define(_createSymbol("bool", TypeClassification.typeSpecifier));
    define(_createSymbol("_Bool", TypeClassification.typeSpecifier));
    define(_createSymbol("_Complex", TypeClassification.typeSpecifier));
    define(_createSymbol("_Decimal32", TypeClassification.typeSpecifier));
    define(_createSymbol("_Decimal64", TypeClassification.typeSpecifier));
    define(_createSymbol("_Decimal128", TypeClassification.typeSpecifier));
    define(_createSymbol("__m128", TypeClassification.typeSpecifier));
    define(_createSymbol("__m128d", TypeClassification.typeSpecifier));
    define(_createSymbol("__m128i", TypeClassification.typeSpecifier));
    define(_createSymbol("__extension__", TypeClassification.typeSpecifier));

    define(_createSymbol("__builtin_va_list", TypeClassification.typeSpecifier));
    define(_createSymbol("__builtin_has_attribute", TypeClassification.typeSpecifier));
    define(_createSymbol("__builtin_speculation_safe_value", TypeClassification.typeSpecifier));
    define(_createSymbol("__builtin_types_compatible_p", TypeClassification.typeSpecifier));
    define(_createSymbol("__builtin_choose_expr", TypeClassification.typeSpecifier));
    define(_createSymbol("__builtin_tgmath", TypeClassification.typeSpecifier));
    define(_createSymbol("__builtin_constant_p", TypeClassification.typeSpecifier));
    define(_createSymbol("__builtin_is_constant_evaluated", TypeClassification.typeSpecifier));
    define(_createSymbol("__builtin_bit_cast", TypeClassification.typeSpecifier));
    define(_createSymbol("__builtin_expect", TypeClassification.typeSpecifier));
    define(_createSymbol("__builtin_expect_with_probability", TypeClassification.typeSpecifier));
    define(_createSymbol("__builtin_trap", TypeClassification.typeSpecifier));
    define(_createSymbol("__builtin_assoc_barrier", TypeClassification.typeSpecifier));
    define(_createSymbol("__builtin_assume_aligned", TypeClassification.typeSpecifier));
    define(_createSymbol("__builtin_LINE", TypeClassification.typeSpecifier));
    define(_createSymbol("__builtin_FUNCTION", TypeClassification.typeSpecifier));
    define(_createSymbol("__builtin_FILE", TypeClassification.typeSpecifier));
    define(_createSymbol("__builtin___clear_cache", TypeClassification.typeSpecifier));
    define(_createSymbol("__builtin_prefetch", TypeClassification.function_));
    define(_createSymbol("__builtin_classify_type", TypeClassification.typeSpecifier));
    define(_createSymbol("__builtin_extend_pointer", TypeClassification.typeSpecifier));
    define(_createSymbol("__builtin_goacc_parlevel_id", TypeClassification.typeSpecifier));
    define(_createSymbol("__builtin_goacc_parlevel_size", TypeClassification.typeSpecifier));

    define(_createSymbol("inline", TypeClassification.functionSpecifier));
    define(_createSymbol("_Noreturn", TypeClassification.functionSpecifier));
    define(_createSymbol("__inline__", TypeClassification.functionSpecifier));

    define(_createSymbol("__cdecl", TypeClassification.functionSpecifier)); // MS
    define(_createSymbol("__clrcall", TypeClassification.functionSpecifier)); // MS
    define(_createSymbol("__stdcall", TypeClassification.functionSpecifier)); // MS
    define(_createSymbol("__fastcall", TypeClassification.functionSpecifier)); // MS
    define(_createSymbol("__thiscall", TypeClassification.functionSpecifier)); // MS
    define(_createSymbol("__vectorcall", TypeClassification.functionSpecifier)); // MS

    define(_createSymbol("_purecall", TypeClassification.typeSpecifier)); // gcc
    define(_createSymbol("_purecall_handler", TypeClassification.typeSpecifier)); // gcc
    define(_createSymbol("_onexit_t", TypeClassification.typeSpecifier)); // gcc
    define(_createSymbol("_locale_t", TypeClassification.typeSpecifier)); // gcc
    define(_createSymbol("_invalid_parameter_handler", TypeClassification.typeSpecifier)); // gcc
    define(_createSymbol("__inline", TypeClassification.typeSpecifier)); // gcc

    define(_createSymbol("__int8", TypeClassification.typeSpecifier)); // gcc
    define(_createSymbol("__int16", TypeClassification.typeSpecifier)); // gcc
    define(_createSymbol("__int32", TypeClassification.typeSpecifier)); // gcc
    define(_createSymbol("__int64", TypeClassification.typeSpecifier)); // gcc
    define(_createSymbol("__int128", TypeClassification.typeSpecifier)); // gcc
    define(_createSymbol("_Float16", TypeClassification.typeSpecifier)); // gcc
    define(_createSymbol("_Float32", TypeClassification.typeSpecifier)); // gcc
    define(_createSymbol("_Float64", TypeClassification.typeSpecifier)); // gcc
    define(_createSymbol("_Float128", TypeClassification.typeSpecifier)); // gcc
    define(_createSymbol("__v8hf", TypeClassification.typeSpecifier)); // gcc
    define(_createSymbol("__bf16", TypeClassification.typeSpecifier)); // gcc
    define(_createSymbol("__v16bf", TypeClassification.typeSpecifier)); // gcc

    define(_createSymbol("__declspec", TypeClassification.functionSpecifier)); // MS
    define(_createSymbol("__attribute__", TypeClassification.functionSpecifier)); // GCC

    define(_createSymbol("alignas", TypeClassification.alignmentSpecifier));
    define(_createSymbol("align", TypeClassification.alignmentSpecifier));
  }

  Symbol _createSymbol(String name, TypeClassification classification) {
    var symbol = Symbol();
    symbol.name = name;
    symbol.classification = {classification};
    symbol.predefined = true;
    return symbol;
  }

  Symbol _createSymbolMulti(String name, List<TypeClassification> classifications) {
    var symbol = Symbol();
    symbol.name = name;
    symbol.classification = Set.from(classifications);
    symbol.predefined = true;
    return symbol;
  }

  void enterScope(Symbol newScope) {
    if (_scopeStack.isNotEmpty && newScope == _scopeStack.last) return;
    _scopeStack.add(newScope);
  }

  void exitScope() {
    _scopeStack.removeLast();
    if (_scopeStack.isEmpty) {
      throw Exception("Cannot exit global scope");
    }
  }

  Symbol? currentScope() {
    if (_scopeStack.isEmpty) return null;
    return _scopeStack.last;
  }

  bool define(Symbol symbol) {
    var scope = currentScope();
    if (scope == null) return false;
    return defineInScope(scope, symbol);
  }

  bool defineInScope(Symbol scope, Symbol symbol) {
    if (scope.members.containsKey(symbol.name)) {
      return false; // Symbol already defined in the current scope
    }
    symbol.parent = scope;
    scope.members[symbol.name] = symbol;
    return true;
  }

  Symbol? resolve(String name, [Symbol? startScope]) {
    if (startScope == null) {
      // Iterate from innermost (top of stack) to outermost (bottom of stack)
      for (var i = _scopeStack.length - 1; i >= 0; i--) {
        var scope = _scopeStack[i];
        var symbol = scope.members[name];
        if (symbol != null) {
          return symbol;
        }
      }
      return null; // Symbol not found
    } else {
      return startScope.members[name];
    }
  }

  Symbol pushBlockScope() {
    var blockScope = Symbol();
    blockScope.name = "block${++_blockCounter}";
    blockScope.classification = {TypeClassification.block};
    blockScope.predefined = true;
    enterScope(blockScope);
    return blockScope;
  }

  void popBlockScope() {
    exitScope();
  }

  @override
  String toString() {
    var result = StringBuffer();
    if (_scopeStack.isNotEmpty) {
      _toStringHelper(result, _scopeStack[0], 0);
    }
    return result.toString();
  }

  void _toStringHelper(StringBuffer sb, Symbol scope, int depth) {
    var indent = '  ' * depth;
    for (var entry in scope.members.entries) {
      var sym = entry.value;
      if (!sym.predefined) {
        sb.writeln('$indent${sym.toString()}');
      }
      // Recursively print nested scopes
      if (sym.classification.contains(TypeClassification.block) ||
          sym.classification.contains(TypeClassification.function_)) {
        _toStringHelper(sb, sym, depth + 1);
      }
    }
  }
}
